frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -1610,6 +1612,37 @@ function ReaderHighlight:getHighlightBookmarkItem()
     end
 end
 
+function ReaderHighlight:determineHighlightInsertionPosition(page, highlight)
+    logger.err("provider", self.ui.document.provider)
+    logger.err("to determine", highlight)
+    if self.view.highlight.saved[page] == nil or #self.view.highlight.saved[page] == 0 then
+        return 1
+    end
+    local page_highlights = self.view.highlight.saved[page]
+    local highlight_beginning = highlight.pos0
+    if type(highlight_beginning) ~= "string" then
+        highlight_beginning.page = page
+    end
+    if self.ui.document.provider == "crengine" then
+        for i, cur_highlight in ipairs(page_highlights) do
+            local order = self.ui.document:compareXPointers(highlight_beginning, cur_highlight.pos0)
+            if order > 0 then
+                return i
```

Comment:
That is true.

Currently, it will assume that they are sorted correctly, if not, it is not really messing up anything from the previous way either.
I want to make migration cheap. really cheap. A possible way is to sort them once when a book is being opened.
Another is to leave them as-is and this will work for all newly read books.
Another is kind of a versioning scheme. If we save these highlights in a file somewhere, we can check for a key, say `sorted` in the table. If it is non-existent then we will migrate them.

Reply:
> Another is kind of a versioning scheme. If we save these highlights in a file somewhere, we can check for a key, say sorted in the table. If it is non-existent then we will migrate them.

That's what we usually do (at the root of the doc settings, no in the highlights table itself), recently in #8616.
I guess for you it would be easy without code duplication: remove all items and put them into a tmp array and just _insertHighlight() back each of them (or just replace the array).

Reply:
I want to do it on document start. I can get `self.view.highlight` on init, but how to save any value there?

Reply:
The same way fixBookmarkSort / bookmarks_sorted_20220106 are handled in readerbookmark.lua (registerPostInitCallback / onSaveSettings) ?

Reply:
Yes.

Reply:
@poire-z

I think I have been able to pull it. Please check whenever you find some time for it.